File: problems/string/PR_72410_신규아이디추천.py
from collections import deque

def solution(new_id):
    unavailable = set("~!@#$%^&*()=+[{]}:?,<>/")

    result = []

    # step 1
    new_id = new_id.lower()
    
    # step 2
    for i in unavailable:
        new_id = new_id.replace(i, '')

    # step 3
    while ".." in new_id:
        new_id = new_id.replace('..', '.')

    result = list(new_id) 
    # step 4
    if result and result[0] == '.':
        result = result[1:]
    if result and result[-1] == '.':
        result.pop()

    # step 5
    if len(result) == 0:
        result.append('a')
    
    # step 6
    if len(result) >= 16:
        result = result[:15]
    if result[-1] == '.':
        result = result[:-1]
    
    # step 7
    while len(result) <= 2:
        result.append(result[-1])

    return ''.join(result)

new_id = "..!@$..[]....a."
print(solution(new_id))

File: problems/string/PR_72410_신규아이디추천.js

function solution(new_id) {
  const answer = new_id
  .toLowerCase()
  .replace(/[^\w-_.]/g, '') // A-Za-z0-9 - _ .을 제외한 애들을 ''로.
  .replace(/\.+/g, '.')     // 한 개 이상의 .이 연속되면 
  .replace(/^\.|\.$/g, '')  // 시작^이 .이거나(\.) 끝$이 .인 경우
  .replace(/^$/, 'a')  // 시작도 끝도 없다. 비어있다.
  .slice(0, 15).replace(/\.$/, '')

  return answer.padEnd(3, answer[answer.length - 1]);
}
